packages/vite/src/node/plugins/reporter.ts

```diff
@@ -115,16 +115,29 @@ export function buildReporterPlugin(config: ResolvedConfig): Plugin {
       for (const id of chunk.moduleIds) {
         const module = this.getModuleInfo(id)
         if (!module) continue
+
+        const hasStaticImport = module.importers.length > 0
+        const hasDynamicImport = module.dynamicImporters.length > 0
+
         // When a dynamic importer shares a chunk with the imported module,
         // warn that the dynamic imported module will not be moved to another chunk (#12850).
-        if (module.importers.length && module.dynamicImporters.length) {
+        if (hasStaticImport && hasDynamicImport) {
+          const warningCondition = (module: string) => {
+            const isExternalDependency = module.includes('/node_modules/')
+            const isInSameChunk = chunk.moduleIds.includes(module)
+            if (!isInSameChunk) return false
+            if (!isExternalDependency) return true
+            return config.build.warnExternalChunkRender
+          }
+
+          const detectedIneffectiveDynamicImport =
+            module.dynamicImporters.some(warningCondition)
```

Comment:
I think we should also make this only a single variable, I'm ok with extracting this out as `detectedIneffectiveDynamicImport` as otherwise the `if` condition would be big. Something like this:
```suggestion
          const detectedIneffectiveDynamicImport = module.dynamicImporters.some(
            (id) => !isInNodeModules(id) && chunk.moduleIds.includes(id),
          )
```

The `isInNodeModules` function come from `utils.ts`.

Moving the `Filter out the ...` comment up above this variable would also be great.

Reply:
it looks more concise